protocol/rep/rep.go

```diff
@@ -227,9 +225,8 @@ func (r *rep) RecvHook(m *mangos.Message) bool {
 	}
 	r.sock.SetSendError(nil)
 	r.backtraceL.Lock()
-	r.backtrace = append(r.backtracebuf[0:0], m.Header...)
+	r.backtrace = append(r.backtrace[0:0], m.Header...)
 	r.backtraceL.Unlock()
-	m.Header = nil
```

Comment:
I'm pretty sure what I was trying to do was gain some caching benefit.  That said, I don't think I achieved success there.

I also think your attempt to filter out duplicates is somewhat naive.  In particular, the request IDs are not globally unique, and you can up misfiltering.  Even the pipe IDs don't help here.  In order to filter out duplicates, you'd have to look at the full stack, and that's probably still not perfect as the route can legitimately change between the sender and receiver.

To be honest, if you have this particular kind of concern where the retry timer is too short, you have two options.  The first is to increase your retry timer.  The second, and probably one to do anyway, is to go ahead and add a unique request ID to your application payload.  ON the rep server, you'd probably want to figure out how to discard duplicates -- by only replying to the lost message once, or (perhaps better) by discarding the incoming request if they arrive while the first one is still being processed.  (Arguably the cooked mode REP socket could do this automatically, if it had a way to uniquely identify the request ID.  It doesn't.  So you'll probably have to do this in RAW mode.)

Frankly, if you're running this in a server that can take a long time to reply, you probably want to be in RAW mode anyway, to avoid blocking other requests.

Once you've sent out a response, you should not discard further request duplicates, since they can arrive due to your response being lost.  In this case, if you've cached the work result & answer (and I hope your service idempotent!), then you can just fire off an immediate reply.

All of this means that I'm no longer sure that its a good idea to proceed with this change.  While it *could* be helpful to you in your particular scenario, I'm afraid that the temptation will be to make use of this in ways that are frankly inappropriate.  Without a unique way to identify the originating requester and request, you can't do automatic dup detection.

(A nice future feature might to be figure out how to add a unique ID to the req message.  However, there are privacy concerns with that approach too, since it would prevent certain kinds of anonymizers from operating.  That said, I'm not sure how big a concern that is in real-world nanomsg/mangos deployments.)